lib/__flake-version-info.nix

```diff
@@ -0,0 +1,17 @@
+# lib overlay to be used by the nixpkgs & nixpkgs/lib flakes
+# to provide meaningful values for `lib.trivial.version` et al..
+#
+# Internal and subject to change, don't use this anywhere else!
+
+self: # from the flake
+
+finalLib: prevLib: # lib overlay
+
+{
+  trivial = prevLib.trivial // {
+    versionSuffix =
+      ".${finalLib.substring 0 8 (self.lastModifiedDate or self.lastModified or "19700101")}.${self.shortRev or "dirty"}";
+    version = finalLib.trivial.release + finalLib.trivial.versionSuffix;
+    revisionWithDefault = default: self.rev or default;
```

Comment:
This is partially duplicated code and is not very re-usable. Instead we can add these to proper `lib.trivial`:
- `lib.trivial.lastModifiedDate` (no `lastModified` since now Flakes always provide `lastModifiedDate`), defaulting to `null`.
- `lib.trivial.revision` (maybe also a `shortRev`, but not necessary since it can easily be computed), defaulting to what `revisionWithDefault` currently defaults to.

Then there's no need to have any implementation here, because it can be moved to the original definitions for `versionSuffix` and `revisionWithDefault`. You will have to remove the `rec` from `lib/trivial.nix` though in order for `version` to be updated correctly.

Then this here can just be

```nix
{
  lastModifiedDate = self.lastModifiedDate;
  revision = self.rev;
}
```

This also allows us to test and document this functionality in `lib/tests`, which I really think should be done.

Reply:
> `trivial`

I don't think this is insignificant or self-evident.

> re-usable

Not sure if that should be a goal here.

> test and document this functionality

Good point, but `lib/tests` can import a file like this to achieve that.

Reply:
> like these to proper lib.trivial:

To make sure I'm not misunderstanding you: you're talking about `lib/trivial.nix` itself, correct?

> no lastModified since now Flakes always provide lastModifiedDate

Good point. Was there for historical reasons only I guess.

> You will have to remove the rec from lib/trivial.nix though in order for version to be updated correctly.

That can be done independently of the other changes you suggested here IMHO.

That said, the downside is that you'll have to touch the entire code affecting `.version-suffix` because that's where the `{lastmodified}.{shortrev}` thing actually comes from in the non-flake case.

> but not necessary since it can easily be computed

Not necessarily because git IIRC determines a length that makes each short rev unique (I think hardcoding a short-rev to e.g. 8 may cause a few collisions already).

Reply:
> you're talking about `lib/trivial.nix` itself, correct?

Yes.



> That said, the downside is that you'll have to touch the entire code affecting `.version-suffix` because that's where the `{lastmodified}.{shortrev}` thing actually comes from in the non-flake case.

No I'm thinking to change the definition of `revisionWithDefault` https://github.com/NixOS/nixpkgs/blob/0a1051cba784efc392695736399c17e51e687e44/lib/trivial.nix#L212-L221 to have an extra `if` condition to use `lib.trivial.revision` if it's not `null`. But I'm not insisting on that, I think it's okay the way this PR is now.



> Not necessarily because git IIRC determines a length that makes each short rev unique (I think hardcoding a short-rev to e.g. 8 may cause a few collisions already).

Ah true, I guess the short revs would still be prone to collisions over longer time though. I'd prefer a sufficiently long constant short rev length that is reasonably collision resistant for the future.

Reply:
>  to have an extra if condition to use lib.trivial.revision if it's not null. But I'm not insisting on that, I think it's okay the way this PR is now.

Hmm I don't think I like that: having an attribute that only works magically when using flakes and that provides a revision despite all the other attributes for that seems unintuitive to me at best.

IMHO redesigning this version code here is another topic and the what we have here is quite an improvement already, so I'd actually prefer the current version for now.